pkg/sbom/diff_nodelist.go

```diff
@@ -0,0 +1,256 @@
+	var removedNodeList NodeList
+	for _, removedNode := range nlDiff.NodesDiff.Removed {
+		removedNodeList.AddNode(removedNode)
+	}
+	for _, removedEdge := range nlDiff.EdgesDiff.Removed {
+		removedNodeList.AddEdge(removedEdge)
+	}
+	removedNodeList.RootElements = append(removedNodeList.RootElements, nlDiff.RootElmementsDiff.Removed...)
+
+	for _, diffNodeRemoved := range nlDiff.NodesDiff.NodeDiff {
+		removedNodeList.AddNode(diffNodeRemoved.Removed)
+	}
+
+	return &removedNodeList
+}
+
+// NodeListDiffNodes represents the differences between two NodeList nodes.
+type NodeListDiffNodes struct {
+	Added    []*Node
+	Removed  []*Node
+	NodeDiff []*NodeDiff
+}
+
+// NodeListDiffEdges represents the differences between two NodeList edges.
+type NodeListDiffEdges struct {
+	Added   []*Edge
+	Removed []*Edge
+}
+
+// NodeListRootElementDiff represents the differences between two NodeLists root elements.
+type NodeListRootElementDiff struct {
+	Added   []string
+	Removed []string
+}
+
+// Diff analyses a NodeList and returns a NodeList populated with all fields
+// that are different in nl2 from nl.
+func (nl *NodeList) Diff(nl2 *NodeList) NodeListDiff {
+	return NodeListDiff{
+		NodesDiff:         nl.diffNodes(nl2, false),
+		EdgesDiff:         nl.diffEdges(nl2),
+		RootElmementsDiff: nl.diffRootElements(nl2),
+	}
+}
+
+// DiffWithStripQulifiers analyses a NodeList and returns a NodeList populated with all fields
+// that are different in nl2 from nl after stripping identifier Quilifiers.
+func (nl *NodeList) DiffWithStripQulifiers(nl2 *NodeList, strip bool) NodeListDiff {
+	return NodeListDiff{
+		NodesDiff:         nl.diffNodes(nl2, strip),
+		EdgesDiff:         nl.diffEdges(nl2),
+		RootElmementsDiff: nl.diffRootElements(nl2),
+	}
+}
+
+// diffRootElements computes the differences in root elements between two NodeLists.
+func (nl *NodeList) diffRootElements(nl2 *NodeList) NodeListRootElementDiff {
+	diff := NodeListRootElementDiff{}
+
+	nlRoots := make(map[string]struct{}) // map to store root elements of nl
+	for _, root := range nl.RootElements {
+		nlRoots[root] = struct{}{}
+	}
+
+	nl2Roots := make(map[string]struct{}) // map to store root elements of nl2
+	for _, root := range nl2.RootElements {
+		nl2Roots[root] = struct{}{}
+	}
+
+	// Compare root elements and populate the diff
+	for root := range nlRoots {
+		if _, exists := nl2Roots[root]; !exists {
+			diff.Removed = append(diff.Removed, root)
+		}
+	}
+	for root := range nl2Roots {
+		if _, exists := nlRoots[root]; !exists {
+			diff.Added = append(diff.Added, root)
+		}
+	}
+
+	sort.Strings(diff.Added)
+	sort.Strings(diff.Removed)
+
+	return diff
+}
+
+func stripQualifiers(url string) string {
+	// Define a regular expression pattern to match everything after "?"
+	pattern := regexp.MustCompile(`\?.*`)
+	// Substitute the matched pattern with an empty string to remove it
+	strippedURL := pattern.ReplaceAllString(url, "")
+	return strippedURL
+}
+
+func stripNodeQualifiers(n *Node) {
```

Comment:
Watch out: This function is mutating the original node's properties as it is working directly on the pointer.

Reply:
removed the hack